dask/array/core.py

```diff
@@ -5651,89 +5651,104 @@ def _vindex_array(x, dict_indexes):
+    )
+    starts = (b[i] for i, b in zip(block_idxs, bounds2))
+    inblock_idxs = np.broadcast_arrays(
+        *(idx - start for idx, start in zip(dict_indexes.values(), starts))
+    )
 
     chunks = [c for i, c in zip(flat_indexes, x.chunks) if i is None]
-    n_chunks, remainder = divmod(len(points), max_chunk_point_dimensions)
+    n_chunks, remainder = divmod(npoints, max_chunk_point_dimensions)
     chunks.insert(
         0,
         (
             (max_chunk_point_dimensions,) * n_chunks
             + ((remainder,) if remainder > 0 else ())
-            if points
+            if npoints > 0
             else (0,)
         ),
     )
     chunks = tuple(chunks)
 
-    if points:
-        per_block = groupby(3, points)
-        per_block = {k: v for k, v in per_block.items() if v}
-
-        other_blocks = list(
-            product(
-                *[
-                    list(range(len(c))) if i is None else [None]
-                    for i, c in zip(flat_indexes, x.chunks)
-                ]
-            )
+    if npoints > 0:
+        other_blocks = product(
+            *[
+                list(range(len(c))) if i is None else [None]
+                for i, c in zip(flat_indexes, x.chunks)
+            ]
         )
 
         full_slices = [slice(None, None) if i is None else None for i in flat_indexes]
 
+        # The output is constructed as a new dimension and then reshaped
+        # So the index of the output point is simply an `arange`
+        outinds = np.arange(npoints).reshape(broadcast_shape)
+        # Which output block is the point going to, and what is the index within that block?
+        outblocks, outblock_idx = np.divmod(outinds, max_chunk_point_dimensions)
+
+        # Now we try to be clever. We need to construct a graph where
+        # if input chunk (0,0) contributes to output chunks 0 and 2 then
+        # (0,0) => (0, 0, 0) and (2, 0, 0).
+        # The following is a groupby over output key by using ravel_multi_index to convert
+        # the (output_block, *input_block) tuple to a unique code.
+        ravel_shape = (
+            n_chunks + 1,
+            *(
+                n
+                for i, n in zip(flat_indexes, x.numblocks, strict=True)
+                if i is not None
+            ),
+        )
+        keys = np.ravel_multi_index([outblocks, *block_idxs], ravel_shape)
+        unique_keys = np.unique(keys)
+
         name = "vindex-slice-" + token
         vindex_merge_name = "vindex-merge-" + token
         dsk = {}
+        mask = np.empty(broadcast_shape, dtype=bool)
         for okey in other_blocks:
             merge_inputs = defaultdict(list)
             merge_indexer = defaultdict(list)
-            for i, key in enumerate(per_block):
+            for i, key in enumerate(unique_keys):
+                outblock, *input_blocks = np.unravel_index(key, ravel_shape)
+                np.equal(keys, key, out=mask)
+                inblock = [_[mask] for _ in inblock_idxs]
```

Comment:
Ideally we'd figure out how to index before broadcasting and broadcast here but that will add a boatload of complexity